benchmark/scenarios/_request_load.py

```diff
@@ -0,0 +1,234 @@
+        self.rate_measurer = RateMeasurer(sample_size)
+        self.max_outstanding = 10 * request_rate
+        # Send requests per second
+        self.loop = LoopingCall.withCount(self._request_and_measure)
+        self.loop.clock = self.reactor
+        # Monitor the status of the scenario
+        self.monitor_loop = LoopingCall(self.check_rate)
+        self.monitor_loop.clock = self.reactor
+
+    def _request_and_measure(self, count):
+        """
+        Update the rate with the current value and send `request_rate`
+        number of new requests.
+
+        :param count: The number of seconds passed since the last time
+            `_request_and_measure` was called.
+        """
+        for i in range(count):
+            self.rate_measurer.update_rate()
+
+        def handle_request_error(result):
+            self.rate_measurer.request_failed(result)
+            write_failure(result)
+
+        for i in range(self.request_rate):
+            d = self.scenario_setup.make_request()
+            self.rate_measurer.request_sent()
+            d.addCallbacks(self.rate_measurer.response_received,
+                           errback=handle_request_error)
+
+    def _fail(self, exception):
+        """
+        Fail the scenario. Stop the monitor loop and throw the
+        error.
+
+        :param exception: `Exception` that caused the failure.
+        """
+        self.monitor_loop.stop()
+        self._maintained.errback(exception)
+
+    def check_rate(self):
+        """
+        Verify that the rate hasn't decreased and that the scenario is
+        not overloaded. A scenario is overloaded if there are too many
+        outstanding requests.
+
+        :raise RequestRateTooLow: if the rate has dropped.
+        :raise RequestOverload: if the scenario is overloaded.
+        """
+        rate = self.rate_measurer.rate()
+        if rate < self.request_rate:
+            self._fail(RequestRateTooLow(rate))
+
+        if self.rate_measurer.outstanding() > self.max_outstanding:
+            self._fail(RequestOverload())
+
+    def start(self):
+        """
+        Runs the setup and starts the scenario
+
+        :return: A Deferred that fires when the desired scenario is
+            established (e.g. that a certain load is being applied).
+        """
+        d = self.scenario_setup.run_setup()
+        d.addCallback(self.run_scenario)
+        return d
+
+    def run_scenario(self, result):
+        """
+        :return: A Deferred that fires when the desired scenario is
+            established (e.g. that a certain load is being applied).
+
+        :raise RequestRateNotReached: if the target rate could not be
+            reached.
+        """
+        self.loop.start(interval=1)
+
+        def reached_target_rate():
+            return self.rate_measurer.rate() >= self.request_rate
+
+        def handle_timeout(failure):
+            failure.trap(CancelledError)
+            raise RequestRateNotReached
```

Comment:
I'm not sure this errback is correct.  Any code that calls `cancel` on `waiting_for_target_rate` will cause `handle_timeout` to be called with a `CancelledError` failure - that includes the `timeout` call below as well as any other timeouts or any other code that decides to use cancellation.

I think the solution is to pass a custom exception to`timeout` and handle that exception in `handle_timeout` instead.